dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/EnrollmentExistingEnrollmentsValidationHook.java

```diff
@@ -0,0 +1,242 @@
+                List<Enrollment> programEnrollments = filterEnrollmentsUserAsAccessTo( actingUser, programInstances );
+
+                // Enrollment(¶4.b.ii) - The error of enrolling more than once is possible only if the imported enrollment
+                // has a state other than CANCELLED...
+                if ( Boolean.TRUE.equals( program.getOnlyEnrollOnce() ) )
+                {
+                    Set<Enrollment> activeOrCompletedEnrollments = programEnrollments.stream()
+                        .filter( programEnrollment -> EnrollmentStatus.ACTIVE == programEnrollment.getStatus()
+                            || EnrollmentStatus.COMPLETED == programEnrollment.getStatus() )
+                        .collect( Collectors.toSet() );
+
+                    if ( !activeOrCompletedEnrollments.isEmpty() )
+                    {
+                        reporter.addError( newReport( TrackerErrorCode.E1016 )
+                            .addArg( trackedEntityInstance )
+                            .addArg( program ) );
+                    }
+                }
+                else if ( EnrollmentStatus.ACTIVE == enrollment.getStatus() )
+                {
+                    // Optimize: this check/filter can be done in fetch above?
+                    Set<Enrollment> activeEnrollments = programEnrollments.stream()
+                        .filter( programEnrollment -> EnrollmentStatus.ACTIVE == programEnrollment.getStatus() )
+                        .collect( Collectors.toSet() );
+
+                    if ( !activeEnrollments.isEmpty() )
+                    {
+                        //Error: TrackedEntityInstance already has an active enrollment in another program...
+                        reporter.addError( newReport( TrackerErrorCode.E1015 )
+                            .addArg( trackedEntityInstance )
+                            .addArg( program ) );
+                    }
+                }
+            }
+
+        }
+
+        return reporter.getReportList();
+    }
+
+    public List<Enrollment> filterEnrollmentsUserAsAccessTo( User actingUser,
+        Iterable<ProgramInstance> programInstances )
+    {
+        List<Enrollment> enrollments = new ArrayList<>();
+
+        for ( ProgramInstance programInstance : programInstances )
+        {
+            if ( programInstance != null && trackerOwnershipManager.hasAccess( actingUser, programInstance ) )
+            {
+                List<String> errors = trackerAccessManager.canRead( actingUser, programInstance, true );
+                if ( errors.isEmpty() )
+                {
+                    enrollments.add( getEnrollmentFromProgramInstance( programInstance ) );
+                }
+                else
+                {
+                    /// ??? // what shall we do here?, this exception seems out of place... shall we report an error instead?
+                    throw new IllegalQueryException( errors.toString() );
```

Comment:
yes, we should return an error. However, checking read access is not always enough. If the operation is enrollment update/delete - we need to check `canWrite`